torch/csrc/api/include/torch/nn/functional/embedding.h

```diff
@@ -1,12 +1,102 @@
 #pragma once
 
+#include <torch/nn/options/embedding.h>
+
 namespace torch {
 namespace nn {
 namespace functional {
 
 inline Tensor one_hot(const Tensor& tensor, int64_t num_classes = -1) {
   return torch::one_hot(tensor, num_classes);
 }
+
+inline Tensor embedding(const Tensor& input, Tensor weight, EmbeddingOptions options) {
+  if (options.padding_idx() != c10::nullopt) {
+    if (*options.padding_idx() > 0) {
+      TORCH_CHECK(*options.padding_idx() < weight.size(0), "Padding_idx must be within num_embeddings");
+    }
+    else if (*options.padding_idx() < 0) {
+      TORCH_CHECK(*options.padding_idx() >= -weight.size(0), "Padding_idx must be within num_embedding");
+      options.padding_idx(weight.size(0) + *options.padding_idx());
+    }
+  } else {
+    options.padding_idx(-1);
+  }
+
+  if (options.max_norm() != c10::nullopt) {
+    torch::NoGradGuard no_grad;
+    torch::embedding_renorm_(weight, input.contiguous(), *options.max_norm(), options.norm_type());
+  }
+  return torch::embedding(weight, input.contiguous(), *options.padding_idx(), options.scale_grad_by_freq(), options.sparse());
+}
+
+inline Tensor embedding_bag(
+    const Tensor& input,
+    Tensor weight,
+    const Tensor& offsets,
+    const EmbeddingBagOptions& options,
+    const Tensor& per_sample_weights) {
```

Comment:
I think we can also add a default constructor for `EmbeddingOptions`, so that we can remove the checks like `TORCH_CHECK((*options).num_embeddings() == embeddings.size(0)` in `Embedding::from_pretrained` and `EmbeddingBag::from_pretrained`

Reply:
that's a good point, but then what should the variables num_embeddings and embedding_dim be set to for the default constructor? They will have to be valid non-zero values otherwise there will be a crash because of the way we are initializing weight in reset.

Reply:
also I was wondering if we should add checks where certain options are not supported under circumstances(instead of just mentioning them in the documentation) like sparse option is not supported when mode = kMax, per_sample_weights only supported for mode = kSum etc

Reply:
> that's a good point, but then what should the variables num_embeddings and embedding_dim be set to for the default constructor? They will have to be valid non-zero values otherwise there will be a crash because of the way we are initializing weight in reset.

That's a good catch! I think we can change `num_embeddings` and `embedding_dim` to be `c10::optional< int64_t >` type, and then add checks to make sure their values are defined in places like the `reset` method (and other places that use `options.num_embeddings()` and `options.embedding_dim()` currently).

Reply:
> also I was wondering if we should add checks where certain options are not supported under circumstances(instead of just mentioning them in the documentation) like sparse option is not supported when mode = kMax, per_sample_weights only supported for mode = kSum etc

Yes I think we currently do have the checks for `sparse option is not supported when mode = kMax` and `per_sample_weights only supported for mode = kSum` in this file. It would be awesome to add the "not-defined" checks for `options.offsets()` and `options.per_sample_weights()` in `EmbeddingBag::reset` as well, and tell the user that these two only take affect in `F::embedding_bag`